docs/proposal-new-output-syntax.md

```diff
@@ -0,0 +1,725 @@
+our conditional reflow prevention mechanism.
+
+For the Cylc 7 graph above to work with the new optional output syntax we just
+need to infer that `a:succeed` and `a:fail` are optional, to avoid `a` ending
+up marked as incomplete both ways.
+
+This example also works on master already:
+```
+# Cylc 7
+[scheduling]
+    [[dependencies]]
+        # If check-d fails, skip d even if c has not finished yet.
+        graph = """
+            a => b => c => d
+            a => check-d => d
+            check-d:fail => !check-d & !d
+        """
+[runtime]
+    [[a,b,c,d]]
+        script = sleep 10
+    [[check-d]]
+        script = false
+```
+The Cylc 8 version still needs a suicide trigger to remove an unsatisfied
+prerequisite:
+```
+# Cylc 8
+[scheduling]
+    [[graph]]
+        # If check-d fails, skip d even if c has not finished yet.
+        R1 = """
+            a => b => c => d
+            a => check-d? => d
+            check-d:fail? => !d
+        """
+[runtime]
+   ...
+```
+This is actually a simple case of the "same task with different dependencies on
+difference branches" problem (see the Appendix below). Cylc 9 will be able to
+handle this properly:
+```
+# Cylc 9
+[scheduling]
+    [[graph]]
+        # If check-d fails, skip d even if c has not finished yet.
+        R1 = """
+            a => b => c
+            a => check-d
+            check-d:fail? => { }  # or just "check-d:fail?"
+            check-d? & c => { d }
+        """
+[runtime]
+   ...
+```
+
+Another example:
+```
+# Cylc 7
+a:x => x1
+a:y => y1
+a:z => z1
+x1 | y1 | z1 => b
+a:x | a:y => ! z1
+a:x | a:z => ! y1
+a:z | a:y => ! x1
+```
+The intention is that `a` completes one (and only one) of the outputs x, y and
+z. Note that the pre-spawning of `b` by Cylc 7 ensures that the workflow will
+stall if none of the outputs are completed, but the suicide triggers will just
+break the workflow if more than one completes.
+
+Using the new syntax this would become:
+```
+# Cylc 8
+a:x? => x1
+a:y? => y1
+a:z? => z1
+x1 | y1 | z1 => b
+
+a(x|y|z)  # a is required to succeed with at least one of x, y, z
+  # OR
+a => b  # spawn b as a prerequisite that needs at least one branch to be satisfied
+```
+
+For backward compatibility all we need to do is, once again, infer that x, y,
+and z are optional outputs (the suicide triggers imply this) and add the
+completion condition. It is probably sufficient to assume the form `a(x|y|z)`
+for all optional output groups.
+
+## Suicide Triggers Still Required at Cylc 8?
+
+Aside from backward compatability we have identified only one remaining problem
+that still requires suicide trigger in Cylc 8: to handle the unwanted extra
+dependencies picked up by a task if it appears on multiple graph branches at
+once (see the Appendix below, and the check-d example above). This will be
+handled cleanly by the dynamic subgraphs planned for Cylc 9; in the meantime
+suicide triggers (and artificial dependencies) provide a decent workaround.
```

Comment:
I may have thought of another?
```
a => b1 & b2? => c
```
We only want `c` to run if both `b1` and `b2` succeed.
We would need to add `b2:fail? => !c` to achieve this.

Reply:
Yes, for this scenario it makes sense to have users get rid of the partially-satisfied prerequisite.